modules/oe_editorial_unpublish/src/Form/NodeUnpublishForm.php

```diff
@@ -0,0 +1,228 @@
+    }
+
+    // Check if the user has a permission to transition to an unpublished state.
+    $workflow = $this->moderationInfo->getWorkflowForEntity($node);
+    $workflow_type = $workflow->getTypePlugin();
+    $unpublished_states = $this->getUnpublishableStates($workflow_type, $node, $account);
+    if (empty($unpublished_states)) {
+      return AccessResult::forbidden($this->t('The user does not have a permission to unpublish the node.'))->addCacheableDependency($node)->addCacheableDependency($workflow)->addCacheableDependency($account);
+    }
+
+    return AccessResult::allowed()->addCacheableDependency($node)->addCacheableDependency($workflow)->addCacheableDependency($account);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getCancelUrl(): Url {
+    return $this->node->toUrl();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getQuestion(): TranslatableMarkup {
+    return $this->t('Are you sure you want to unpublish the node %label?', [
+      '%label' => $this->node->label(),
+    ]);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getConfirmText(): TranslatableMarkup {
+    return $this->t('Unpublish');
+  }
+
+  /**
+   * Returns the states that are available to unpublish.
+   *
+   * @param \Drupal\workflows\WorkflowTypeInterface $workflow_type
+   *   The workflow type plugin.
+   * @param \Drupal\node\NodeInterface $node
+   *   The node that is beind unpublished.
+   * @param \Drupal\core\Session\AccountInterface $account
+   *   A user account to check permissions for, defaults to the current user.
+   *
+   * @return array
+   *   An array of states keyed by the state id.
+   */
+  protected function getUnpublishableStates(WorkflowTypeInterface $workflow_type, NodeInterface $node, AccountInterface $account = NULL): array {
+    $account = $account ?? $this->currentUser();
+    // Gather a list of unpublishable_states.
+    $available_states = $workflow_type->getStates();
+    $unpublishable_states = [];
+    foreach ($available_states as $state) {
+      if (!$state->isPublishedState() && $state->isDefaultRevisionState()) {
+        $unpublishable_states[$state->id()] = $state;
+      }
+    }
+
+    // Gather a list of states to which the node can transition to.
+    $current_state = $node->moderation_state->value;
+    $available_transitions = $workflow_type->getTransitionsForState($current_state);
+    $transitionable_states = [];
+    foreach ($available_transitions as $transition) {
+      $transitionable_states[$transition->to()->id()] = $transition->to();
+    }
+    $unpublishable_states = array_intersect_key($unpublishable_states, $transitionable_states);
+
+    foreach (array_keys($unpublishable_states) as $state_id) {
+      $transition_id = $workflow_type->getTransitionFromStateToState($node->moderation_state->value, $state_id);
+      if (!$account->hasPermission('use oe_corporate_workflow transition ' . $transition_id->id())) {
```

Comment:
You can put that comment message here for checking the permission.

Reply:
Done.